### ISSUE TITLE
[Regression] ResourceLoader.LoadThreadedRequest(scenePath) stuck at 0.5, infinite loop.

### ISSUE THREAD

**Issue by conde2** [godotengine/godot#99128]
### Tested versions

NOTReproducible in:
v4.3 stable for Windows

Reprodcible in:
v4.3 stable for MacOs
v4.4.dev4.mono.official [36e6207bb] (Windows and MacOS)

For MacOS and Windows

### System information

Godot v4.4.dev4.mono - Windows 10.0.19045 - Multi-window, 1 monitor - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Ti (NVIDIA; 32.0.15.6094) - AMD Ryzen 7 3700X 8-Core Processor (16 threads)

### Issue description

When using 
```
ResourceLoader.LoadThreadedRequest(scenePath);
```

This happened before and looks like its still happening:
https://github.com/godotengine/godot/issues/92844#issuecomment-2167849384

NOTE:
I'm calling the render server when loading TileSets, but I'm doing a RenderingServer.ForceSync();

I'm getting stuck at 0.5 progress and the progress never changes, maybe a recursive load? Or a race condition, I don't know, my debugger points here:

![image](https://github.com/user-attachments/assets/504983f5-ffdb-4f7a-a277-8f48f2c21cea)

### Steps to reproduce

I'm loading the instance like this:

In MacOS it doesnt work at all it is always stuck (thats why we did the load in the same thread), on windows there is a race condition, so sometimes it works, sometime it doesnt.

```C#
public partial class Main2 : Node2D
{
    public override void _Ready()
    {
        // Construct the path to the scene file
        var scenePath = $"res://main.tscn";

        var error = ResourceLoader.LoadThreadedRequest(scenePath);
        GD.PrintErr($"Loading scene: {scenePath} {error}");
        var percentageReport = new Godot.Collections.Array();

        double percentage;
        ResourceLoader.ThreadLoadStatus loadingStatus;
        do
        {
            loadingStatus = ResourceLoader.LoadThreadedGetStatus(scenePath, percentageReport);
            if (loadingStatus == ResourceLoader.ThreadLoadStatus.Failed)
            {
                GD.PrintErr($"Failed to load scene at path: {scenePath}");
                return;
            }

            percentage = percentageReport[0].As<double>();
            GD.PrintErr($"Loading scene: loading: {loadingStatus} scene: {scenePath} percentage: {percentage} {percentageReport}");
            RenderingServer.ForceSync();
        } while (loadingStatus == ResourceLoader.ThreadLoadStatus.InProgress);

        GD.PrintErr($"Finished loading scene: {scenePath} {loadingStatus} {percentage}");
    }
}
```
            

### Minimal reproduction project (MRP)

[bug-thread.zip](https://github.com/user-attachments/files/17846554/bug-thread.zip)

**Comment by conde2** [comment]
@RandomShaper I managed to create a MRP:

[bug-thread.zip](https://github.com/user-attachments/files/17846522/bug-thread.zip)


To reproduce just hit the play project multiple times, at some point it will hang and be stuck in the infinite loop.

**Comment by conde2** [comment]
@RandomShaper, @akien-mga I still can reproduce it in 4.4dev7 even after [#99539](https://github.com/godotengine/godot/pull/99539)

Godot version:
v4.4.dev7.mono.official [46c8f8c5c]

Godot v4.4.dev7.mono - Windows 10 (build 19045) - Multi-window, 1 monitor - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Ti (NVIDIA; 32.0.15.6094) - AMD Ryzen 7 3700X 8-Core Processor (16 threads)

**Comment by TCROC** [comment]
On Linux everything works fine for me.  On Android however, I start to run into some interesting things.

1. When loading via ``ResourceLoader.LoadThreadedRequest(scenePath)``, it loads to 50% and then instantly switches to the other scene which is fascinating.
2. When loading via ``ResourceLoader.LoadThreadedRequest(scenePath, useSubThreads: true);`` it loads to 75% before switching on 2 of our 3 scenes.  On the 3rd scene, it hangs forever at 0%.

I have no idea if this is related, but I found the 50% thing interesting when loading linearly and the freeze with subthreads.  I will have to put together a separate mrp so you guys can test.  In summary:

```
ResourceLoader.LoadThreadedRequest(scenePath)
```

10% -> 23% -> 35% -> 50% -> switch scene

```
ResourceLoader.LoadThreadedRequest(scenePath, useSubThreads: true);
```

0% -> 75% -> switch scene

on all except 1 which hangs.

Tbh, the first one likely isn't an issue.  Its hard to know which assets will take longer to load than others, so a clean 0% -> 100% is hard to achieve.  But the hang is definitely a bug.

**Comment by takdew** [comment]
Issue still happens at random, with no clear reproduction steps. I'm trying to load a `Node3D` `PackedScene` using `ResourceLoader.LoadThreadedRequest(scenePath)`. Progress is stuck at `0.5` while polling in a `_Process` method.

Godot v4.4.beta1.mono - Windows 10 (build 19045) - Single-window, 1 monitor - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 4070 (NVIDIA; 32.0.15.6094) - Intel(R) Core(TM) i7-14700K (28 threads)

**Comment by MrModez** [comment]
Godot v4.4.beta2.mono - the problem still occurs...

30% of times ResourceLoader.LoadThreadedRequest gets stuck at 50%

**Comment by blewis-web** [comment]
Godot v4.4.stable.mono - still happening. I agree that this should be a high priority issue.

On my actual (proprietary) project (where I need to load a large scene via a button), I'm seeing the problem more of the time when not calling `RenderingServer.ForceSync()` at all:
60% of the time ResourceLoader.LoadThreadedRequest is getting stuck at 50% (Stuck: 30 times, Completed Normally: 20 times).

When calling `RenderingServer.ForceSync()` immediately after each call to `ResourceLoader.LoadThreadedGetStatus()` in `_Update()`, it appears that my odds improve significantly:
12% of the time ResourceLoader.LoadThreadedRequest is getting stuck at 50% (Stuck: 6 times, Completed Normally: 44 times).

Godot v4.4.stable.mono - Windows 11 (build 22631) - Multi-window, 2 monitors - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 4090 (NVIDIA; 32.0.15.6636) - Intel(R) Core(TM) i9-14900KF (32 threads)

**Comment by bclewi** [comment]
I'm able to reproduce the issue in 4.4 stable (but less frequently than I expected).
Update: Also able to reproduce in `4.4.1.stable.mono`

I was having difficulty with the two links to the MRP above (and it was hard for me to tell what was going on - it would freeze or even crash the editor). I modified/improved it a bit to remove those distractions and now I can see the regression clearly when it happens.

[bug-thread-improved.zip](https://github.com/user-attachments/files/19147185/bug-thread-improved.zip)

<Details>

<Summary>Details (Including screenshots and MRP Change List)</Summary>



Screenshot from when it is stuck:
![Image](https://github.com/user-attachments/assets/561a5eca-7f83-45a5-8fd5-409641125cc4)

Screenshot from when it completes loading normally:
![Image](https://github.com/user-attachments/assets/bf0ed37e-feec-4e90-85ed-32a2e95743fc)

### Change List:
- Main2.cs
  - Rewrote the script to use `_Process` (as recommended in the documentation).
  - Changed logging so that `GD.PrintErr` is only used for problems (`GD.Print` is used for normal debug info).
  - Changed logging so that messages only print once (as it was causing a message overflow issue in the original).
- SomeComponent.tscn
  - Clicked "Access as Unique Name" on `TextureProgressBar` and `Timer` since the C# script was using `GetNode<T>("%NodeName")`

</Details>

Godot v4.4.stable.mono - Windows 11 (build 22631) - Multi-window, 2 monitors - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 4060 (NVIDIA; 32.0.15.6094) - 13th Gen Intel(R) Core(TM) i5-13600K (20 threads)

**Comment by TCROC** [comment]
I fixed an issue just today where Godot freezes on exit in 4.4. I expect it's in earlier versions as well. It became more obvious when updating to LLVM 19. I'll be putting in a PR either tomorrow or Monday. Details on what I found can be seen here:

https://chat.godotengine.org/channel/devel?msg=BiNxQdeH7uD2s6umC

I suspect it's a very similar reason that is causing this issue.

**Comment by blewis-web** [comment]
Not sure how much this helps, but while debugging, when I click "Break All" in Visual Studio when reproducing this issue (in my project and in the MRP - https://github.com/user-attachments/files/19147185/bug-thread-improved.zip), it always stops on this line of code:
![Image](https://github.com/user-attachments/assets/1dd4476f-7f98-4576-b479-1b6bbbc93e94)

Call Stack
> [External Code]
GodotSharp.dll!Godot.Bridge.ScriptManagerBridge.GetOrCreateScriptBridgeForType(System.Type scriptType, Godot.NativeInterop.godot_ref* outScript) Line 574	C#
 	GodotSharp.dll!Godot.Bridge.ScriptManagerBridge.GetOrCreateScriptBridgeForPath(Godot.NativeInterop.godot_string* scriptPath, Godot.NativeInterop.godot_ref* outScript) Line 569	C#

Godot v4.4.stable.mono - Windows 11 (build 22631) - Multi-window, 2 monitors - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 4090 (NVIDIA; 32.0.15.6636) - Intel(R) Core(TM) i9-14900KF (32 threads)

**Comment by jvdnbus** [comment]
After pulling my hair out why `LoadThreadedGetStatus` was returning InvalidResource (after requesting it properly), even though the scene works just fine separately and with `ResourceLoader.Load<PackedScene>`, I stumbled upon these issues here. As of now (4.4.1.stable.mono) this feature seems completely useless. If someone could figure out why this is happening and a fix, it would be greatly appreciated ❤

**Comment by dougVanny** [comment]
I believe I just stumbled upon what I believe is same issue as this. In my case I am running a bigger chunk of code in a separate thread, in which I load and instantiate many packedScenes, either through direct Exports or through ResourceLoader.LoadThreadedRequest, before adding them all back to the scene tree through a deferred call. Randomly, when this thread runs, other operations that are happening in the main thread end up deadlocked, which effectively makes my whole game freeze

The line where the freeze actually occurs is random but consistent. When I press play, it will always freeze at the at the same line of my code. If I make a change completely unrelated to that line, however, such as commenting out a method call somewhere else, the freeze will be caused by a completely different line of code, if it will happen at all. By debugging, I see that I am always getting stuck at the _scriptTypeBiMap.ReadWriteLock of the ScriptManagerBridge class

I have experimented replacing every ResourceLoader.LoadThreadedRequest call by ResourceLoader.Load both in main thread code and in my separate thread. This caused the freezing to stop, but given the random nature of the issue I have a hard time saying that this solved it since I am still loading resources from a separate thread anyway, and that is bound to pass through that same ReedWriteLock that is causing the deadlocks. I am considering simply rewiring my systems so that, instead of calling ResourceLoader.LoadThreadedRequest, I'll manually create a new Thread and call ResourceLoader.Load from it, but right now I'm hesitant about if I really should make my game rely on asynchronous resource loading or not.

**Comment by dougVanny** [comment]
I have given this issue some more thought and I have a theory on how to fix it, but unfortunately I'm not equipped to compile and test godot

From what I understand, as described on #99798, the source of this issue is the fact that, while loading a script and while the lock is acquired, dependencies might have to be loaded as well. If the loading is made from a worker thread, a new thread will be created and go through the same lock, and therefore the deadlock will happen as the previous thread already had locked it. Debugging on VS, this is exactly what I saw, it tries to load a class while loading a resource through `ResourceLoader.LoadThreadedRequest`, and then tries to load it's base class, which it fails to due to the lock

The fix proposal above was to remove the `godot_sharp_internal_script_load` from the lock, which might have fixed the issue, but instead it was preferred to go with #99539, which keeps that call in the lock, but instead used an actual ReadWrite lock so that, in theory, the deadlock couldn't happen anymore

The issue, however, is that the actual behavior of the lock remained exactly the same! All lock calls were replaced by `ReaderWriterLockSlim.EnterUpgradeableReadLock` and additional `ReaderWriterLockSlim.EnterWriteLock` calls were added. An UpgradeableReadLock can run in parallel with other ReadLocks, but not with other UpgradeableReadLocks, so despite all the apparent Read locking, there is never more than one thread reading at the same time. When `godot_sharp_internal_script_load` is called the thread is not under a write lock, but it is under an upgradeable lock, and so the new thread to load any dependencies gets stuck

Considering all that, my proposal would be to mix both solutions, keeping `godot_sharp_internal_script_load` inside a read lock but not in a upgradeable lock. According to the [ReaderWriterLockSlim documentation](https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-threading-readerwriterlockslim#upgrade-and-downgrade-locks), it is possible to downgrade a lock back to a normal ReadLock simply by calling `ReaderWriterLockSlim.EnterReadLock`. Releasing the lock is still be done with `ReaderWriterLockSlim.ExitUpgradeableReadLock` so `ReaderWriterLockSlim.ExitReadLock` is not required

I do not have enough knowledge of the `ScriptManagerBridge` class to know just where Read locks actually need to be upgradeable or not, but I am under the impression that `godot_sharp_internal_script_load` could be done in a Read only lock by downgrading it. Right after it, the code my either call `GetOrCreateScriptBridgeForType` or `AddScriptBridgeCore`, but each of these start by already acquiring a lock, so I'm lead to believe that there won't be a problem if another Upgradeable lock is acquired before those methods are actually executed. If this is a problem, then we might need to rethink if loading dependencies should actually create a new thread or not

I hope these findings help shed some light on how to fix this issue :)

**Comment by RandomShaper** [comment]
@dougVanny Oh, if there's a section not needing an upgradeable lock, a plain read one makes sense indeed. Also, it may be the case that in some cases it's not necessary to have a read lock upgraded to write, being possible instead to read-lock, unlock and finally write-lock. However, I didn't want to break the critical section because I couldn't be sure I wouldn't be introducing any race conditions by doing so.

**Comment by calvinchd** [comment]
I stumbled upon a clue or a possible fix. I can load consistently after moving `CreateScriptBridgeForType` out of the upgradeable read lock in `ScriptManagerBridge.GetOrCreateScriptBridgeForType` like so:

```cs
private static unsafe void GetOrCreateScriptBridgeForType(Type scriptType, godot_ref* outScript)
{
    _scriptTypeBiMap.ReadWriteLock.EnterUpgradeableReadLock();
    try
    {
        if (_scriptTypeBiMap.TryGetScriptPtr(scriptType, out IntPtr scriptPtr))
        {
            // Use existing
            NativeFuncs.godotsharp_ref_new_from_ref_counted_ptr(out *outScript, scriptPtr);
            return;
        }
    }
    finally
    {
        _scriptTypeBiMap.ReadWriteLock.ExitUpgradeableReadLock();
    }
    // This path is slower, but it's only executed for the first instantiation of the type
    CreateScriptBridgeForType(scriptType, outScript);
}
```
Edit made on commit 80a3d205f1ad22e779a64921fb56d62b893881ae

This may introduce a race condition between exiting the upgradeable lock and entering the write lock in `CreateScriptBridgeForType` and I am unsure if `NativeFuncs.godotsharp_internal_reload_registered_script` in `CreateScriptBridgeForType` needs to be in a lock.

**Comment by conde2** [comment]
@calvinchd this is the same fix applied in this pull request that was reject:

https://github.com/godotengine/godot/pull/99798

**Comment by calvinchd** [comment]
> [@calvinchd](https://github.com/calvinchd) this is the same fix applied in this pull request that was reject:
> 
> [#99798](https://github.com/godotengine/godot/pull/99798)

My bad, just ignore my comment

**Comment by conde2** [comment]
@TCROC I believe he "fixed" it using a similar approach in [this pull request](https://github.com/godotengine/godot/pull/99798). However, that PR wasn't merged because the Godot team chose to proceed with [this alternative PR](https://github.com/godotengine/godot/pull/99539), which unfortunately does not resolve the issue.

So we still have the problem, and the approach he mentioned was considered risk by the godot team, as stated in the pull request I mentioned.

**Comment by TCROC** [comment]
Thank you for the clarification!  I'll be curious to test out these PRs.  I have this issue as well and as a result, just don't use ``LoadThreadedRequest`` with multiple threads.  Just a single background thread.

**Comment by calvinchd** [comment]
@TCROC I had the issue after #99539. #99798 is more complete and less hacky than what I attempted

**Comment by TCROC** [comment]
Thanks for the heads up @calvinchd !  Did https://github.com/godotengine/godot/pull/99798 fix it for you as well?  I'll test it out on mine and if it fixes it, we will likely want to reevaluate reopening the PR.

**Comment by calvinchd** [comment]
@TCROC These changes from #99798 fixed the issue for me

[99798_changes.txt](https://github.com/user-attachments/files/20007634/99798_changes.txt)

Moving `NativeFuncs.godotsharp_internal_reload_registered_script` out of the lock in `GetOrCreateScriptBridgeForType` was enough to fix it for me.
Moving `NativeFuncs.godotsharp_internal_reload_registered_script` out of the lock in `GetOrLoadOrCreateScriptForType.GetPathOtherwiseGetOrCreateScript` didn't change anything on my end.

**Comment by gvvim** [comment]
As a temporary workaround while this gets patched, in C# you can do: 
```
await Task.Run(() =>
	{
		ResourceLoader.Load("res://Your/resource.res", cacheMode: ResourceLoader.CacheMode.Reuse);
	}
);
```
Once at game startup to cache it in advance without blocking the main thread. Subsequent loads start at 0.81 and don't get stuck in an infinite loop.

**Comment by dougVanny** [comment]
Is there currently any consensus or anyone investigating the appropriate way to correctly fix this bug? I saw that 4.5 beta 1 was just released so I was wondering if there are any hopes of including this in the final 4.5 release 

It's just my personal opinion but I find this a pretty severe bug, it's preventing me from doing any asynchronous resource loading which means freezing my game for a few frames every time I have to load something heavy. Were my project in a more advanced state, this bug would be enough to stop me from updating Godot until it was fixed

I see it's in the 4.x Release Blockers project but it's status is simply Bad, and there is an actual status for Release Blockers so it isn't clear to me if I should expect a fix or not

I wish I could do more to help fix this. Thanks to everyone who is putting effort into this bug ^^

**Comment by raulsntos** [comment]
If the closed PR (https://github.com/godotengine/godot/pull/99798) fixes the issue you all are getting and doesn't introduce other issues, then we can reopen it and merge it. It was closed because we didn't have enough confidence it wouldn't introduce other bugs, and it initially seemed the alternative PR would fix this bug too.

@TCROC Did you end up testing https://github.com/godotengine/godot/pull/99798? If so, did that PR fix the issue for you?

@dougVanny You said you are unable to compile Godot, would you be willing to test a custom build that includes https://github.com/godotengine/godot/pull/99798? If so, what platform does the editor need to target? And if you need to test this in an exported game, what platforms do the templates need to target?

In general, the more people test and confirm that https://github.com/godotengine/godot/pull/99798 fixes the issue without introducing other issues, the more confident we'll be to merge the PR. I'm currently unable to reproduce this issue, so testing and reviews from other users and contributors is much appreciated.

**Comment by TCROC** [comment]
@raulstnos I have not tested it yet. We've recently started testing game with our users so we've been prioritizing the bugs introduced by us in our game logic itself. But we are due for an engine update so I'll include that PR in our next built here soon for a test

**Comment by dougVanny** [comment]
> [@dougVanny](https://github.com/dougVanny) You said you are unable to compile Godot, would you be willing to test a custom build that includes [#99798](https://github.com/godotengine/godot/pull/99798)? If so, what platform does the editor need to target? And if you need to test this in an exported game, what platforms do the templates need to target?

@raulsntos Definitely! I'm currently working with 4.4.1 in Windows x86_64 but I could also test in a 4.5 build if you find it better to test it like that. Currently I'm just exporting to Windows as well

**Comment by raulsntos** [comment]
@dougVanny Here you go:

Editor | Templates
-|-
[Windows x86_64](https://github.com/raulsntos/godot/releases/download/pr99798/Godot_v4.5-pr99798_mono_win64.zip) | [Windows x86_64](https://github.com/raulsntos/godot/releases/download/pr99798/Godot_v4.5-pr99798_mono_export_templates.tpz)

It's a build from `master` at commit 4e6ac9bccc7868432b9d8a009d34860a2e6f3393  with https://github.com/godotengine/godot/pull/99798  rebased and included.

**Comment by dougVanny** [comment]
@raulsntos I think I need the nupkg to be able to compile my project. I'm getting the following error

```
Could not resolve SDK "Godot.NET.Sdk". Exactly one of the probing messages below indicates why we could not resolve the SDK. Investigate and resolve that message to correctly specify the SDK.
  SDK resolver "Microsoft.DotNet.MSBuildWorkloadSdkResolver" returned null.
  Unable to find package Godot.NET.Sdk with version (= 4.5.0-pr.99798)
  - Found 124 version(s) in nuget.org [ Nearest version: 4.5.0-dev.5 ]
  - Found 0 version(s) in Microsoft Visual Studio Offline Packages
  
  MSB4236: The SDK 'Godot.NET.Sdk/4.5.0-pr.99798' specified could not be found.
```

**Comment by raulsntos** [comment]
I think all the packages you need should be in the **Editor** compressed archive, under the `GodotSharp/Tools/nupkgs` directory. Put them in a [custom NuGet source feed](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_with_dotnet.html#nuget-packages) or extract them to your NuGet packages directory (`%userprofile%\.nuget\packages.` in Windows).

**Comment by dougVanny** [comment]
@raulsntos After attempting it several times, I could no longer reproduce the bug with your build. I could reproduce it quite easily on 4.4.1 so I think #99798 solves it for me :)

**Comment by preslavnpetrov** [comment]
Hey folks, just caught up with the discussion here. @raulsntos reached out to me to reopen that PR 👍.

Thanks for testing!

**Comment by TCROC** [comment]
Got around to testing this.  I see there are merge conflicts that I'm not really sure how to resolve.  @preslavnpetrov , @ me when you rebase it and I'll test it out in my fork

**Comment by raulsntos** [comment]
@TCROC This is the branch I used to make the custom build: https://github.com/raulsntos/godot/tree/dotnet/rebase-99798

**Comment by tehKaiN** [comment]
I can also confirm that the fix from #99798 solved the problems for me.

**Comment by bclewi** [comment]
@raulsntos, I can confirm that using your [custom Windows build](https://github.com/raulsntos/godot/releases/download/pr99798/Godot_v4.5-pr99798_mono_win64.zip) for #99798 fixes the issue for me in my (very easy to test) MRP [bug-thread-improved.zip](https://github.com/user-attachments/files/19147185/bug-thread-improved.zip). I tried running it hundreds of times and it never got stuck! In contrast, it only took ~20 times or less for me to reproduce it again today on `4.4.0` and `4.4.1`.
